### PR TITLE
Time shift difference

### PR DESCRIPTION
**Pull request by betodealmeida**
I added a new control to allow comparing comparing the main time series with time shifts. In addition to showing the time shift as individual lines, it's now possible to compute absolute and relative differences.

Current behavior, will be the default option in the selector ("individual lines"):

screen shot 2018-06-11 at 4 03 21 pm

Absolute difference:

screen shot 2018-06-11 at 4 03 28 pm

Relative difference, computed as `(main - time_shift) / time_shift` (for a "1 week" shift" this means that `100` means the value today is double the value last week, and `-50` means that the value today is half of what it was last week):

screen shot 2018-06-11 at 4 03 37 pm

### CODE REVIEW

**Comment by betodealmeida** [comment]
> I think this is effectively the same as a superior Period Ratio isn't it? If so we should db migrate to translate the slices.

I'm not familiar with the period ratio, and it doesn't seem to map 1:1 to the comparison types. Should I create a migration script only for those cases that can be mapped?

> I know one difference in the backend is that Period Ratio runs 2 queries and merges them. So I could look at the past 28 days and compare with last year's same period. I'm not sure if Time Shift can do that or it would query/show more than 28 days (which may be ok). What do you think?

How I would that with the period ratio? For time shift you would select the last 28 days and then use "1 year" as the time shift.

**Comment by vylc** [comment]
Some updates to the AA section to improve usability
screen shot 2018-06-12 at 3 33 09 pm

File: superset/assets/src/explore/controls.jsx [hunk 1]
```diff
@@ -1562,24 +1562,6 @@ export const controls = {
     description: t('Compute the contribution to the total'),
   },
 
-  num_period_compare: {
```
Comment (mistercrunch) [inline superset/assets/src/explore/controls.jsx]: If deprecating `num_period_compare` and `period_ratio_type`, do we need a db migration for existing charts that have been saved using that feature?

**Comment by mistercrunch** [comment]
LGTM apart from the db migration needed comment.

**Comment by betodealmeida** [comment]
@mistercrunch, I've added a migration script. Here's a dashboard that was saved before the migration:

screen shot 2018-07-06 at 6 45 00 pm

Note that:

- granularity is `month`;
- `num_period_compare` is 24 (months);
- `period_ratio_type` is "growth" (ie, `new/old - 1`).

After the migration:

screen shot 2018-07-06 at 8 04 02 pm

Now:

- `time_compare` is "2 years" (though it doesn't show in the dropdown since it's free form);
- `comparison_type` is "percentage" (same as "growth", just renamed).

**Comment by mistercrunch** [comment]
LGTM

**Comment by williaster** [comment]
@betodealmeida this broke my local build, the db migration fails

File: superset/assets/package.json [hunk 2]
```diff
@@ -68,6 +68,7 @@
     "geojson-extent": "^0.3.2",
     "geolib": "^2.0.24",
     "immutable": "^3.8.2",
+    "is-react": "^1.1.1",
```
Comment (williaster) [inline superset/assets/package.json]: @betodealmeida @mistercrunch was it necessary to add this dependency for functionality that is already provided by React natively? 

from the [source](https://github.com/treyhuffine/is-react/blob/master/index.js#L32): 
`React.isValidElement(typeElement)`

Comment (williaster) [reply superset/assets/package.json]: FYI this added react 16 as a direct dependency (not a peer dependency), which means we have 2 versions of react floating around. it'd be great to get rid of this.

Comment (betodealmeida) [reply superset/assets/package.json]: Let me take a look.

Comment (betodealmeida) [reply superset/assets/package.json]: Fixed in https://github.com/apache/incubator-superset/pull/5478.

File: superset/viz.py [hunk 3]
```diff
@@ -1219,24 +1205,50 @@ def run_extra_queries(self):
...
 
-        return chart_data
+                # remove leading/trailing NaNs from the time shift difference
+                diff = diff[diff.first_valid_index():diff.last_valid_index()]
+
+                chart_data.extend(
+                    self.to_series(
+                        diff, classed='time-shift-{}'.format(i), title_suffix=label))
+
+        return sorted(chart_data, key=lambda x: tuple(x['key']))
```
[the remark this answers is not in the document]
Comment (betodealmeida) [reply superset/viz.py]: Taking a look at this.

[the remark this answers is not in the document]
Comment (betodealmeida) [reply superset/viz.py]: IIRC this is needed in order to color the groups in different timeshifts with the same color (eg, make sure "boys" is red in the main series as well as all offsets).

**Pull request merged**